src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,220 @@
+                ->update($this->configuration['table_name'])
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id');
+            $this->executeQuery($sql, [
+                ':id' => $doctrineEnvelope['id'],
+                ':delivered_at' => (\DateTime::createFromFormat('U.u', microtime(true)))->format('Y-m-d H:i:s.u'),
+            ]);
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Throwable $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+    }
+
+    public function reject($id): bool
+    {
+        return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+    }
+
+    private function executeQuery(string $sql, array $parameters = [])
+    {
+        $stmt = null;
+        try {
+            $stmt = $this->driverConnection->prepare($sql);
+            $stmt->execute($parameters);
+        } catch (TableNotFoundException $e) {
+            // create table
+            if (!$this->driverConnection->isTransactionActive() && $this->configuration['auto_setup']) {
+                $this->setup();
+            }
+            // statement not prepared ? SQLite throw on exception on prepare if the table does not exist
+            if (null === $stmt) {
+                $stmt = $this->driverConnection->prepare($sql);
+            }
+            $stmt->execute($parameters);
+        }
+
+        return $stmt;
+    }
+
+    private function setup(): void
+    {
+        $synchronizer = new SingleDatabaseSynchronizer($this->driverConnection);
+        $synchronizer->updateSchema($this->getSchema(), true);
+    }
+
+    private function getSchema(): Schema
+    {
+        $schema = new Schema();
+        $table = $schema->createTable($this->configuration['table_name']);
+        $table->addColumn('id', Type::BIGINT)
+            ->setAutoincrement(true)
+            ->setNotnull(true);
+        $table->addColumn('body', Type::TEXT)
+            ->setNotnull(true);
+        $table->addColumn('headers', Type::JSON)
+            ->setNotnull(true);
+        $table->addColumn('queue', Type::STRING)
+            ->setLength(32)
+            ->setNotnull(true);
+        $table->addColumn('created_at', Type::DATETIME)
+            ->setLength(3)
+            ->setNotnull(true);
+        $table->addColumn('available_at', Type::DATETIME)
+            ->setLength(3)
+            ->setNotnull(true);
+        $table->addColumn('delivered_at', Type::DATETIME)
+            ->setLength(3)
+            ->setNotnull(false);
+        $table->setPrimaryKey(['id']);
+        $table->addIndex(['queue']);
+        $table->addIndex(['available_at']);
+        $table->addIndex(['delivered_at']);
```

Comment:
I'm not sure if these are the best indexes or not - maybe a DB expert can chime in and help. For reference, the query will look something like this - 3 WHERE clauses and an ORDER BY:

> SELECT m.* FROM messenger_messages m WHERE (m.delivered_at is null OR m.delivered_at < '2019-03-03 00:01:02') AND (m.available_at <= '2019-03-03 00:01:02') AND (m.queue = 'messages') ORDER BY available_at ASC LIMIT 1

Reply:
I'm not a DB expert either. Just added those indexes because these columns are used in `WHERE` and `ORDER BY` clauses as you point it :)